### PR TITLE
Patch Durable Extension version check

### PR DESCRIPTION
**Pull request by davidmrdavid**
Currently, when the host checks the durable extension version, it makes two errors:
1) It assumes all apps using extension bundles are using DF Extension V1.x
2) It assumes all apps using C# isolated are using DF Extension V1.x because it cannot find `extensions.json` in the `/bin` directory

This PR fixes this issue in the following ways:

For (1), given that only bundles V2 onwards is supported in Functions V4, we assume accordingly DF Extension V2.x is used.
For (2), we fix the location where `extensions.csproj` is looked to also include its location in isolated .NET apps.

**Implication:** During the SyncTriggers call, the Host will provide different metadata depending on whether DF Extension V2.x or V1.x is used. When testing target-based scaling, the Host was previously providing DF Extension V1.x-style metadata for all out of process apps. Now, we should be receiving the expected SyncTriggers metadata.

### Pull request checklist

* [x] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [ ] My changes **should not** be added to the release notes for the next release
    * [x] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] My changes **do not** require diagnostic events changes
    * Otherwise: I have added/updated all related diagnostic events and their documentation (Documentation issue linked to PR)
* [x] I have added all required tests (Unit tests, E2E tests)

### Additional information

Additional PR information

### CODE REVIEW

**Comment by davidmrdavid** [comment]
FYI @pragnagopa, I suppose we can update the images after merge. Thank you

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: The flow here should match what is used for the extension probing, using the same order of precedence:

1 - `bin` folder
2 - script root
3 - System path (`.azurefunctions`) which is the constant `AzureFunctionsSystemDirectoryName` inside of the script root path.

You can reference the logic here:
- https://github.com/Azure/azure-functions-host/blob/66731fa6715f02610c3d0842c56488e540e695e3/src/WebJobs.Script/DependencyInjection/ScriptStartupTypeLocator.cs#L128-L154

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: Should we be getting the actual version from the metadata instead of hardcoding this here?

**Review by fabiocav — CHANGES_REQUESTED** [review]

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (davidmrdavid) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: @fabiocav: you mean from the extension bundles version range? I thought of that, but worried that parsing the version range via regex might be dangerous and more error prone than it's worth, so I opted for just going with bundles V1 not being supported for Functions V4. Happy to be challenged on this though.

Is there code in the host that can already parse the bundles range? If so, I can re-use that

Comment (fabiocav) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: Here, do you care about the bundle or DF major version?

Comment (davidmrdavid) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: DF major version. Maybe I misunderstood your suggestion. Can you please elaborate?

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (davidmrdavid) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: Ah, good to know. I've added the check for the system path by referencing that logic. Let me know if that looks good to you

File: test/WebJobs.Script.Tests.Integration/Management/FunctionsSyncManagerTests.cs
Comment (davidmrdavid) [inline test/WebJobs.Script.Tests.Integration/Management/FunctionsSyncManagerTests.cs:477]: This test assumed that bundles meant we're using DF V1, so I changed it accordingly.

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: Why are we duplicating this logic from ScriptStartupTypeLocator? They need to be kept in sync right - we want to use the same logic. Can't we factor the existing code into a shared helper that you can simply call here?

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs:549]: What is the plan for addressing https://github.com/Azure/azure-functions-host/issues/5390? We really need to get this logic out of the host. Durable is the only extension that has custom logic on this path.

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs:554]: The new behavior will be on by default. Will this break existing apps that are currently running and potentially working around the existing behavior via config?

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (davidmrdavid) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs:549]: No plan yet, I was not aware of https://github.com/Azure/azure-functions-host/issues/5390 until today. I've replied to that thread to continue that conversation, but I would like to keep that separate from this PR. That's beyond the scope of this bug fix, I think.

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (davidmrdavid) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs:554]: Our understanding, from discussing with @pragnagopa, is that it should not break existing apps. The consumer of SyncTriggers is the scale controller and ScaleController V2 can work both with the old buggy behavior (in that case it compensates for the missing SyncTriggers metadata) and with the new behavior introduced in this PR (in which case it doesn't need to compensate at all).

We've also added the opt-out flag `FeatureFlagEnableLegacyDurableVersionCheck` as a safety mechanism to opt out of this change in case we're wrong about this assessment.

File: src/WebJobs.Script/DependencyInjection/ScriptStartupTypeLocator.cs
Comment (davidmrdavid) [inline src/WebJobs.Script/DependencyInjection/ScriptStartupTypeLocator.cs]: Just refactoring this so it can be -reused as per: https://github.com/Azure/azure-functions-host/pull/9331#discussion_r1232543085

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (davidmrdavid) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: Good idea, I refactor the logic in ScriptStartupTypeLocator in my latest commit to do just this.

Comment (davidmrdavid) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: You can see the relevant diff here: https://github.com/Azure/azure-functions-host/pull/9331/commits/8b1fcd15ee4fa1cf287fcd16105df9f541e7d6b9

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (davidmrdavid) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs:554]: To clarify: this bug fix is motivated by ScV3 which does not have this compensating logic. That's how we discovered this bug

**Review by mathewc — APPROVED** [review]

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (fabiocav) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: That would be available (as the actual version) with the extensions resolved by the the bundle. I don't think we need to actually block on that, but lookin at the actual extension type would give you the exact major version. We already have code in the script startup type locator to parse the extensions once you get what is resolved by the bundles. 

Again, not a blocker.

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (fabiocav) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: Thanks, that's what I was hoping for with my comment [here](https://github.com/Azure/azure-functions-host/pull/9331#discussion_r1230306366)

File: src/WebJobs.Script/DependencyInjection/ScriptStartupTypeLocator.cs
Comment (fabiocav) [inline src/WebJobs.Script/DependencyInjection/ScriptStartupTypeLocator.cs]: nit (super nit): Can we rename this to `ResolveExtensionsMetadataPath`?

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: Questioning the use of the static method on `ScriptStartupTypeLocator` from here. The alternative would be to move it to `Utility` (or some other more appropriate type).

OK without the change too.

**Review by fabiocav — APPROVED** [review]

File: src/WebJobs.Script/DependencyInjection/ScriptStartupTypeLocator.cs
Comment (davidmrdavid) [reply src/WebJobs.Script/DependencyInjection/ScriptStartupTypeLocator.cs]: How about `TryResolveExtensionsMetadataPath` :) . I generalized this method a little bit now that it's been moved to `Utility.cs` so that it returns a `bool` indicating whether or not it's search was successful. Happy to be challenged on this

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (davidmrdavid) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: I had similar thoughts but I was trying to minimize changes unless explicitly requested :). Moved to `Utility`

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (davidmrdavid) [inline src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs:452]: comment requested by @mathewc

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (davidmrdavid) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: That would be a better approach, so I'd be happy to refine this logic later. However, as it's not a blocker, and given that Host v4 already fails fast when bundles V1 is loaded (just validated this locally), I think I'd rather keep this change as-is for simplicity and expediency's sake.

If all goes according to plan, we should be removing this DF specific logic from the Host anyways, so hopefully this is a short term solution for a codepath that is "soon" to be removed :)

**Review by fabiocav — APPROVED** [review]

**Pull request merged**
